effect/jvm/src/main/scala/scalaz/effect/RTS.scala

```diff
@@ -0,0 +1,1058 @@
+import scalaz.data.Maybe
+
+/**
+ * This trait provides a high-performance implementation of a runtime system for
+ * the `IO` monad on the JVM.
+ */
+trait RTS {
+  import RTS._
+
+  /**
+   * Effectfully and synchronously interprets an `IO[A]`, either throwing an
+   * error, running forever, or producing an `A`.
+   */
+  def unsafePerformIO[A](io: IO[A]): A = extract(tryUnsafePerformIO(io))
+
+  /**
+   * Effectfully interprets an `IO`, blocking if necessary to obtain the result.
+   */
+  final def tryUnsafePerformIO[A](io: IO[A]): Throwable \/ A = {
+    // TODO: Optimize — this is slow and inefficient
+    var result: Try[A] = null
+    lazy val lock = new ReentrantLock()
+    lazy val done = lock.newCondition()
+
+    val context = new FiberContext[A](this, defaultHandler)
+
+    context.evalSync(io)
+
+    (context.register { (r: Try[A]) =>
+      try {
+        lock.lock()
+
+        result = r
+
+        done.signal()
+      } finally lock.unlock()
+    }) match {
+      case AsyncReturn.Now(v) =>
+        result = v.asInstanceOf[Try[A]]
+
+      case _ =>
+        while (result == null) try {
+          lock.lock()
+          done.await()
+        } finally lock.unlock()
+    }
+
+    result
+  }
+
+  /**
+   * The default handler for unhandled exceptions in the main fiber, and any
+   * fibers it forks that recursively inherit the handler.
+   */
+  val defaultHandler: Throwable => IO[Unit] =
+    (t: Throwable) => IO.sync(t.printStackTrace())
+
+  val threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors().max(2))
+
+  /**
+   * This determines the maximum number of resumptions placed on the stack
+   * before a fiber is shifted over to a new thread to prevent stack overflow.
+   */
+  val MaxResumptionDepth = 10
+
+  /**
+   * Determines the maximum number of operations executed by a fiber before
+   * yielding to other fibers.
+   *
+   * FIXME: Replace this entirely with the new scheme.
+   */
+  val YieldMaxOpCount = 1024
+
+  lazy val scheduledExecutor = Executors.newScheduledThreadPool(1)
+
+  final def submit[A](block: => A): Unit = {
+    threadPool.submit(new Runnable {
+      def run: Unit = { block; () }
+    })
+
+    ()
+  }
+
+  final def schedule[A](block: => A, duration: Duration): AsyncReturn[Unit] =
+    if (duration == Duration.Zero) {
+      submit(block)
+
+      AsyncReturn.later[Unit]
+    } else {
+      val future = scheduledExecutor.schedule(
+        new Runnable {
+          def run: Unit = { submit(block) }
+        }, duration.toNanos, TimeUnit.NANOSECONDS)
+
+      AsyncReturn.maybeLater { (t: Throwable) => future.cancel(true); () }
+    }
+}
+
+private object RTS {
+  // Utility function to avoid catching truly fatal exceptions. Do not allocate
+  // memory here since this would defeat the point of checking for OOME.
+  def nonFatal(t: Throwable): Boolean =
+    if (t.isInstanceOf[InternalError]) false
+    else if (t.isInstanceOf[OutOfMemoryError]) false
+    else true
```

Comment:
could be `!t.isInstanceOf[InternalError] && !t.isInstanceOf[OutOfMemoryError]`

Reply:
or even !(t.isInstanceOf[InternalError] || t.isInstanceOf[OutOfMemoryError]), which is a little bit closer to how i'd talk in english

Reply:
Fixed. 👍